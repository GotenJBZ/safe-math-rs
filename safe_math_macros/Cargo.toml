[package]
name = "safe_math_macros"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[features]
derive = ["num-traits"]

[dependencies]
syn = { version = "2", features = ["full", "fold"] }
quote = "1"
proc-macro2 = "1"
num-traits = { version = "0.2.19", optional = true }
