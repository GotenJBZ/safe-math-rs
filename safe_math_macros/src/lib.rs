use proc_macro::TokenStream;
use quote::quote;
use syn::{BinOp, Expr, ExprBinary, ItemFn, parse_macro_input};

#[proc_macro_attribute]
pub fn safe_math(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let mut input_fn = parse_macro_input!(item as ItemFn);
    let orig_block = *input_fn.block;
    let new_block = rewrite_block(orig_block);
    input_fn.block = Box::new(new_block);
    TokenStream::from(quote! { #input_fn })
}

fn rewrite_block(block: syn::Block) -> syn::Block {
    use syn::fold::{self, Fold};
    struct MathRewriter;
    impl Fold for MathRewriter {
        fn fold_expr(&mut self, expr: Expr) -> Expr {
            match expr {
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::Add(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { ::safe_math::safe_add(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::Sub(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { ::safe_math::safe_sub(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::Mul(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { ::safe_math::safe_mul(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::Div(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { ::safe_math::safe_div(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::Rem(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { ::safe_math::safe_rem(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::AddAssign(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { #left = ::safe_math::safe_add(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::SubAssign(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { #left = ::safe_math::safe_sub(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::MulAssign(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { #left = ::safe_math::safe_mul(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::DivAssign(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { #left = ::safe_math::safe_div(#left, #right)? }
                }
                Expr::Binary(ExprBinary {
                    left,
                    op: BinOp::RemAssign(_),
                    right,
                    ..
                }) => {
                    let left = self.fold_expr(*left);
                    let right = self.fold_expr(*right);
                    syn::parse_quote! { #left = ::safe_math::safe_rem(#left, #right)? }
                }
                _ => fold::fold_expr(self, expr),
            }
        }
    }
    MathRewriter.fold_block(block)
}
