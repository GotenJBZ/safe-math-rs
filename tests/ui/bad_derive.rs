use safe_math_macros::SafeMathOps;

// Duplicate operation "add"
#[derive(SafeMathOps)]
#[SafeMathOps(add, add)]
struct Foo(i32);

// Unknown operation "unknown"
#[derive(SafeMathOps)]
#[SafeMathOps(unknown)]
struct Bar(i32);

// Missing list entirely
#[derive(SafeMathOps)]
struct Baz(i32);

fn main() {} 