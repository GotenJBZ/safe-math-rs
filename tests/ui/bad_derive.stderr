error: Duplicate operation 'add' in `#[SafeMathOps]` attribute. Each operation should be listed only once.
 --> tests/ui/bad_derive.rs:5:20
  |
5 | #[SafeMathOps(add, add)]
  |                    ^^^

error: Unknown operation 'unknown' in `#[SafeMathOps]` attribute. Supported operations are: add, sub, mul, div, rem.
  --> tests/ui/bad_derive.rs:10:15
   |
10 | #[SafeMathOps(unknown)]
   |               ^^^^^^^

error: `#[SafeMathOps]` requires at least one operation, e.g. `#[SafeMathOps(add, sub)]`
  --> tests/ui/bad_derive.rs:15:1
   |
15 | struct Baz(i32);
   | ^^^^^^^^^^^^^^^^
